pkg/services/ngalert/schedule/schedule.go

```diff
@@ -52,6 +52,15 @@ type RecordingWriter interface {
 	Write(ctx context.Context, name string, t time.Time, frames data.Frames, orgID int64, extraLabels map[string]string) error
 }
 
+// AlertRuleStopReasonProvider is an interface for determining the reason why an alert rule was stopped.
+type AlertRuleStopReasonProvider interface {
+	// FindReason returns two values:
+	// 1. The first value is the reason for stopping the alert rule (error type).
```

Comment:
I kept the first value an error type because that's what is currently passed via the context and [CancelCauseFunc](https://github.com/grafana/grafana/blob/d24e7c126ddb107515e9a36ab1285c05cc639fe2/pkg/util/contextutil.go#L25)